crates/parser/src/grammar/items/use_item.rs

```diff
@@ -70,24 +70,30 @@ fn use_tree(p: &mut Parser<'_>, top_level: bool) {
                 // main balanced `{}`
                 p.err_and_bump(msg);
             }
-            return;
+            return false;
         }
     }
     m.complete(p, USE_TREE);
+    true
 }
 
+pub(super) const USE_TREE_LIST_RECOVERY_SET: TokenSet =
+    TokenSet::new(&[T![;], T![,], T![.], T![ident]]);
+
 // test use_tree_list
 // use {a, b, c};
 pub(crate) fn use_tree_list(p: &mut Parser<'_>) {
     assert!(p.at(T!['{']));
     let m = p.start();
-    p.bump(T!['{']);
-    while !p.at(EOF) && !p.at(T!['}']) {
-        use_tree(p, false);
-        if !p.at(T!['}']) {
-            p.expect(T![,]);
-        }
-    }
-    p.expect(T!['}']);
+
+    // test_err use_tree_list_err_recovery
+    // use {a;
+    // use b;
+    // struct T;
+    // fn test() {}
+    delimited(p, T!['{'], T!['}'], T![,], USE_TREE_LIST_RECOVERY_SET, |p: &mut Parser<'_>| {
```

Comment:
`USE_TREE_LIST_RECOVERY_SET` is wrong here, that param in `delimited` is not for recovery but for the first set. So we want a `USE_TREE_LIST_FIRST_SET` that contains the tokens that use tree may start with.